### ISSUE TITLE
Pinned memory fails to be unregisterd

### ISSUE THREAD

**Issue by nv-dlasalle** [dmlc/dgl#3800]
## 🐛 Bug
If you allocate and free pinned memory in a loop in DGL currently, you will get:
```
terminate called after throwing an instance of 'c10::CUDAError'
  what():  CUDA error: resource already mapped
```
When that same memory region gets reallocated.

## To Reproduce

Steps to reproduce the behavior if you modify the current `examples/pytorch/graphsage/multi_gpu_node_classification.py` to allocate pinned memory inside of the inference loop:
```
def shared_tensor(*shape, device, q):
    rank = dist.get_rank()
    world_size = dist.get_world_size()
    # make sure q is only used for one operation at a time
    dist.barrier()
    if rank == 0:
        y = torch.zeros(
            *shape, device=device)
        for i in range(1, world_size):
            q.put(y)
    else:
        y = q.get()
    dist.barrier()
    pin_memory_inplace(y) # <-- error occurs here
    return y


class SAGE(nn.Module):
    def __init__(self, in_feats, n_hidden, n_classes):
        super().__init__()
        self.layers = nn.ModuleList()
        self.layers.append(dglnn.SAGEConv(in_feats, n_hidden, 'mean'))
        self.layers.append(dglnn.SAGEConv(n_hidden, n_hidden, 'mean'))
        self.layers.append(dglnn.SAGEConv(n_hidden, n_classes, 'mean'))
        self.dropout = nn.Dropout(0.5)
        self.n_hidden = n_hidden
        self.n_classes = n_classes

    def forward(self, blocks, x):
        h = x
        for l, (layer, block) in enumerate(zip(self.layers, blocks)):
            h = layer(block, h)
            if l != len(self.layers) - 1:
                h = F.relu(h)
                h = self.dropout(h)
        return h

    def inference(self, g, device, batch_size,
                  num_workers, buffer_device, q):
        sampler = dgl.dataloading.MultiLayerFullNeighborSampler(1)

        # manually partition the IDs
        dataloader = dgl.dataloading.NodeDataLoader(
                g, torch.arange(g.num_nodes(), device=device), sampler, device=device,
                batch_size=batch_size, shuffle=False, drop_last=False, num_workers=num_workers,
                persistent_workers=(num_workers > 0), use_ddp=True, use_uva=True)

        feat = g.ndata['feat']
        for l, layer in enumerate(self.layers):
            y = shared_tensor(
                    g.num_nodes(), self.n_hidden if l != len(self.layers) - 1 else self.n_classes,
                    device=buffer_device, q=q)

            for input_nodes, output_nodes, blocks in dataloader:
                x = gather_pinned_tensor_rows(feat, input_nodes)
                h = layer(blocks[0], x)
                if l != len(self.layers) - 1:
                    h = F.relu(h)
                    h = self.dropout(h)
                # make use of the fact that output nodes is a continuous range
                y[output_nodes] = h.to(buffer_device)
            # make sure all GPUs are done writing to 'y'
            dist.barrier()
            feat = y
        # return only nodes in the set
        return y
```

The last layer will attempt to re-use an overlapping memory allocation of that of the first layer. Upon attempting to pin the new allocation, the CUDA error is returned.

## Expected behavior

When a pinned tensor is free'd, we would expect the page-locked memory to also be released.

## Environment

 - DGL Version (e.g., 1.0): 0.8
 - Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3): 1.10
 - OS (e.g., Linux): ubunt 18.04
 - How you installed DGL (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.8
 - CUDA/cuDNN version (if applicable): 11.3
 - GPU models and configuration (e.g. V100): V100 16 SXM2
 - Any other relevant information:

**Comment by nv-dlasalle** [comment]
Strangely this only occurs with multiple processes. With one process I don't get this error, and if I accumulate all pinned tensors in a list so they are not freed, I also do not get it.

**Comment by davidmin7** [comment]
I think we need to set an internal flag in DGL's NDArray upon calling `pin_memory_()` and also define `__del__` which will check the flag and call `unpin_memory_()` (There could be a more elegant way of doing this, but this is what I can think of now).

It is interesting that the error explicitly occurs only at the multi-processing situation, but for the sake of correctness it is better to be explicitly unpinned regardless of the single/multi-processing cases.

**Comment by yaox12** [comment]
Agree with David Min. The tensor to be pinned here is a PyTorch tensor and its deconstructor will never unpin the tensor. A solution is to manually unpin it at the end of the loop. I will send you the version I modified via slack. @nv-dlasalle

**Comment by nv-dlasalle** [comment]
@yaox12 I've investigated this more and tested why your change was working, and it seems when we pass a tensor through a pipe, we need to synchronize all processes before pinning it, or this error occurs later on when pinning a second tensor.

```
def shared_tensor(*shape, device, q):
    rank = dist.get_rank()
    world_size = dist.get_world_size()
    if rank == 0:
        y = torch.zeros(
            *shape, device=device)
        for i in range(1, world_size):
            q.put(y)
    else:
        y = q.get()
    dist.barrier() # this must be after all calls to `put` and `get`, but before pinning.
    pin_memory_inplace(y)
    return y
```

Un-registering the memory appears to be a separate but related issue.

**Comment by jermainewang** [comment]
Here is my proposal:
* When setting `use_uva=True` and the input graph is not pinned, `DataLoader` will pin the graph for users inside `__init__`. When setting `use_uva=True` and the input graph is already pinned, `DataLoader` will not pinned the graph for users. In this way, users can control when to pin and unpin it outside of `DataLoader`.
* Inside the destructor of the internal C++ graph data class, we will call `unpin` if the the `is_pinned` flag is true. This makes sure that when a pinned graph object is destroyed, the memory page will be released. I think the PyTorch tensor may has a similar mechanism -- users can pin a tensor but dont need to unpin it manually. Note that this will only help graph but not feature tensor. For the later, we probably need to upstream the request to PyTorch.

**Comment by nv-dlasalle** [comment]
I think our model needs changing. Imagine the case where we have two graphs, one larger one for "pretraining", and a smaller one for final training. The users code looks something like:
```
def pretrain(...):
  pre_train_dataloader(g_pretrain, ..., use_uva=True)
  pre_test_dataloader(g_pretrain, ..., use_uva=True)
  for epoch in range(n_pre_epochs):
    for ... in pre_train_dataloader:
      ...
  for ... in pre_test_dataloader:
    ...

def train(...):
  train_dataloader(g_train, ..., use_uva=True)
  test_dataloader(g_train, ..., use_uva=True)
  for epoch in range(n_train_epochs):
    for ... in train_dataloader:
      ...
  for ... in test_dataloader:
    ...
```
In this case, we need to unpin the feature data associated with `g_pretrain` since it will be freed when the pretraining is complete, and we don't want to be re-allocated while still being pinned. However, we can't unpin it in the destructor of the dataloader, as the test dataloader still needs it to be pinned when the training dataloader is no longer needed.

I think we need to pin memory outside of the dataloader, and have the dataloader complain to the user if they have failed to do so:
```
def pretrain(...):
  with g_pretrain.pinned() as g_pinned:
    pre_train_dataloader(g_pinned, ..., use_uva=True)
    pre_test_dataloader(g_pinned, ..., use_uva=True)
    for epoch in range(n_pre_epochs):
      for ... in pre_train_dataloader:
        ...
    for ... in pre_test_dataloader:
      ...

def train(...):
  with g_train.pinned() as g_pinned:
    train_dataloader(g_pinned, ..., use_uva=True)
    test_dataloader(g_pinned, ..., use_uva=True)
    for epoch in range(n_train_epochs):
      for ... in train_dataloader:
        ...
    for ... in test_dataloader:
      ...
```

**Comment by nv-dlasalle** [comment]
Importing @davidmin7 's comment from another PR:



>> The problem is for array allocated outside of DGL, we can only modify the Deleter DGL will call, but not the one PyTorch/MXNET/Tensorflow will call.

> Is it possible to do a deleter chaining? For example:

```
// curr_deleter is the deleter which will be called by default
original_deleter = curr_deleter;
curr_deleter = unpin;
```

> and inside the unpin() function we do something like:
```
void unpin() {
    // Do unpinning jobs here
    ...
    // Chain to the original deleter
    original_deleter();
}
```

> In this case, wouldn't the unpinning job be done during the destructor stage even if the tensor was allocated by PyTorch? The deleter function is passed as a pointer in dlpack so I think we can create a linked list of the destructors and insert the unpinning stage before the 'original' deleter is called. 2 cents of my thoughts on this issue.

**Comment by yaox12** [comment]
> Is it possible to do a deleter chaining?

I think it's already the case in DGL. First we will call `DLPackDeleter`, inside which we will call the tensor's original deleter.

**Comment by nv-dlasalle** [comment]
This isn't fixed for feature data allocated by pytorch but pinned by DGL (as is done in the dataloader).

**Comment by yaox12** [comment]
> > Is it possible to do a deleter chaining?
> 
> I think it's already the case in DGL. First we will call `DLPackDeleter`, inside which we will call the tensor's original deleter.

When investigating https://github.com/dmlc/dgl/pull/4061, I realize we cannot do a deleter chaining for now because when we pin the tensor allocated by pytorch but not through tensoradapter. For example, in
```python
F.to_dgl_nd(pytorch_tensor).pin_memory_()
```
the only chance we have to give the tensor a mark is in `F.to_dgl_nd` or `NDArray::FromDLPack` more concretely. During pinning, what we can manipulate is the underlying `DLTensor`.  That's to say we cannot give an external tensor a `pinned_by_dgl` flag or an unpin deleter.

**Comment by jermainewang** [comment]
Closed this issue because PR#4135 has fixed it for most of the cases. For new problems, please open a new issue ticket.
